Introduce task queues, and throttling performance timeline tasksource

gterzian opened the pull request:
<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #19997 (github issue number if applicable).

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/21388)
<!-- Reviewable:end -->

gterzian committed:
introduce task-queues, and throttling the performance-timeline task-source, in script and worker threads.

queue

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/script_thread.rs, components/script/lib.rs, components/script/task_queue.rs
 * @kichjang: components/script/script_thread.rs, components/script/lib.rs, components/script/task_queue.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits include an empty title element (`<title></title>`). Consider adding appropriate metadata.
* These commits modify **unsafe code**. Please review it carefully!

gterzian commented:
@jdm this is sort of a first draft on the "task-queue" concept, let me know what you think...

gterzian commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 9b886e66f5f4e64254f97b019d7bd95e458c2b3d with merge 551a6fcb432a3a227d29129d9030d87c02748993...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/12184), [android-x86](http://build.servo.org/builders/android-x86/builds/298), [arm32](http://build.servo.org/builders/arm32/builds/12382), [arm64](http://build.servo.org/builders/arm64/builds/12426), [linux-dev](http://build.servo.org/builders/linux-dev/builds/12476), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/9399), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/9398), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/12518), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/4955), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/4959), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/9226), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/9268), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/5028), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/4972), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/7875)
State: approved= try=True

gterzian commented:
Last commit is an attempt at #20723

gterzian commented on components/script/dom/dedicatedworkerglobalscope.rs:
```diff
@@ -70,20 +71,61 @@ impl<'a> Drop for AutoWorkerReset<'a> {
+    fn recv_tasks(&self) -> Vec<WorkerScriptMsg> {
+        self.receiver.try_iter().map(|msg|{
+            let (linked_worker, script_msg) = msg;
+            let _ar = AutoWorkerReset::new(&*self.dedicated_scope, linked_worker);
```
I think this is actually not going to work, since we should only do the `AutoWorkerReset` 'right before' handling a particular event...

gterzian commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 07375b777d76bf9940eaa8f2be5bff87de450084 with merge ab43f9a95f130d03d7d254981251fa1f001133ac...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/9471)

gterzian commented:
```
{"status": "TIMEOUT", "group": "default", "message": "Test timed out", "stack": null, "subtest": "Throttling the performance timeline task queue.", "test": "/_mozilla/mozilla/task_queue_throttling.any.worker.html", "line": 165581, "action": "test_result", "expected": "PASS"}
{"status": "ERROR", "group": "default", "message": "Error in remote: uncaught exception: unknown (can't convert to string)", "stack": null, "subtest": null, "test": "/_mozilla/mozilla/task_queue_throttling.any.worker.html", "line": 165582, "action": "test_result", "expected": "OK"}
```

It seems the test I wrote is somewhat unstable, it failed on only one box... I'll look into it

gterzian commented:
@bors-servo try (got rid of the `setTimeout` in the test)

bors-servo commented:
:hourglass: Trying commit f641af7e40cf9320edee238819ae906a0b915f87 with merge b1cb6ab4ac3bc27be9168109de939bd261bbc122...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/9473)

gterzian commented:
@bors-servo try (changed the expectations of the test so that it doesn't depend on a specific amount of throttling, since when tasks themself are 'slow' to come in, as seems to happen in linux-rel-css, the performance timeline will be throttled yet we can't be sure how many other tasks will have been executed before)

bors-servo commented:
:hourglass: Trying commit a98e5b6035da1b8825d7e2141708d57ccff3ee9a with merge e4f97da86bf9760fe5171de8fba5085f220b6b86...

jdm commented:
@asajeffrey You might be interested in these changes.

gterzian commented:
(added some comments)

gterzian commented:
I've added some comments, I actually think this takes us closer to spec compliance with the worker processing model, the worker event-loop, without the proposed changes, stops once there are no more events to received, while per the spec it should keep running until the `close` flag as been set. See Step 29 of https://html.spec.whatwg.org/multipage/workers.html#worker-processing-model:terminate-a-worker-3 and https://html.spec.whatwg.org/multipage/workers.html#worker-event-loop

gterzian committed:
introduce a generic worker event-loop

gterzian commented:
I've made an attempt at getting to 'one event-loop' code for both workers, yet wasn't able to get it to compile. It seems hard to have something like `WorkerEventLoop<T: WorkerEventLoopMethods>` were `WorkerEventLoopMethods` is `WorkerEventLoopMethods<A, B, C>`

You could take a look https://github.com/servo/servo/commit/fe4a39a1179cc598060bbb66422307a82ef3c6b1

gterzian commented:
Ok so the last commit is a "generic worker event-loop" because it was hurting a bit to see those two event-loops doing essentially the same thing. It might be taking it a bit too far with the generics, anyway, for your consideration...

gterzian commented on components/script/task_queue.rs:
```diff
@@ -151,11 +152,25 @@ impl<T: QueuedTaskConversion> TaskQueue<T> {
     }
 
+    // Time an iteration of the event-loop.
+    fn time_event_loop(&self) {
```
This seems to actually be a duplicate of https://github.com/servo/servo/blob/d34403047e806fa6c8c2468946f64429622ec434/components/script/script_thread.rs#L1249

I think I'll remove it...

gterzian commented:
Cleaned-up a few things left-over from the rebase, also got rid of that `allow(unrooted_must_root)`, it is now 'more ready for review'...

gterzian commented on components/script/dom/abstractworkerglobalscope.rs:
```diff
@@ -55,12 +62,98 @@ impl<T: JSTraceable + DomObject + 'static> ScriptChan for WorkerThreadWorkerChan
+    fn handle_event(&self, event: Self::Event);
+    fn handle_worker_post_event(&self, worker: &TrustedWorkerAddress);
+    fn worker_global_scope(&self) -> &WorkerGlobalScope;
+    fn global_scope(&self) -> &GlobalScope;
```
thinking of replace this method with one called `perform_microtask_checkpoint`, since that's all that is being done with the global scope as part of running the event-loop...

gterzian commented on components/script/dom/abstractworkerglobalscope.rs:
```diff
@@ -55,12 +62,98 @@ impl<T: JSTraceable + DomObject + 'static> ScriptChan for WorkerThreadWorkerChan
+    fn task_queue(&self) -> &TaskQueue<Self::WorkerMsg>;
+    fn handle_event(&self, event: Self::Event);
+    fn handle_worker_post_event(&self, worker: &TrustedWorkerAddress);
+    fn worker_global_scope(&self) -> &WorkerGlobalScope;
```
We could also replace this with two methods:
1. `should_stop`, which internally would check the `closing` flag of the `WorkerGlobalScope`, and
2. `devtools_sender`, which would return the devtools sender from the `WorkerGlobalScope`

bors-servo commented:
:umbrella: The latest upstream changes (presumably #21519) made this pull request unmergeable. Please resolve the merge conflicts.

jdm reviewed (changes requested):
This implementation is really readable once I got my head around how it works. Nice job! I really appreciate the deduplication of the worker event loops, too.

gterzian commented on components/script/dom/dedicatedworkerglobalscope.rs:
```diff
@@ -293,14 +347,44 @@ impl DedicatedWorkerGlobalScope {
-            panic!("unexpected select result!")
+        let event = {
+            if ret == worker_handle.id() {
+                MixedMessage::FromWorker(self.task_queue.take_tasks().recv().unwrap())
```
I think this will go away with https://github.com/servo/servo/pull/21325

gterzian commented on components/script/dom/serviceworkerglobalscope.rs:
```diff
@@ -24,22 +24,77 @@ use js::jsapi::{JSAutoCompartment, JSContext, JS_AddInterruptCallback};
+            ServiceWorkerScriptMsg::CommonWorker(worker_script_msg) => worker_script_msg,
+            _ => return None,
+        };
+        let script_msg = match worker_script_msg {
```
Do you mean doing both operations in one match? If so, done...

jdm replied:
I actually meant the two pieces of code that end up performing this series of matches (see previous review comment).

gterzian commented on components/script/task_queue.rs:
```diff
@@ -0,0 +1,201 @@
+
+        for msg in to_be_throttled {
+            // Categorize tasks per task queue.
+            let (worker, category, boxed, pipeline_id) = msg.into_queued_task().unwrap();
```
I changed it to a match with `unreachable`. 

I think there is a contract that a message whose category is non-priority, hence was included in `to_be_throttled` should implement `into_queued_task` that returns `Some`, otherwise it would point to a logical error either in the `filter` above, or in the implementation of `QueuedTaskConversion` for that message type. 

We still need an option, because not all variants of the message represent something that could be converted into a queued task(this was basically the main difficulty of getting to a generic task queue, since the variant containing a `CommonScriptMsg::Task` is wrapped into different enum with different structures).

gterzian commented on components/script/task_queue.rs:
```diff
@@ -0,0 +1,201 @@
+            // Categorize tasks per task queue.
+            let (worker, category, boxed, pipeline_id) = msg.into_queued_task().unwrap();
+            let task_source = match category {
+                ScriptThreadEventCategory::PerformanceTimelineTask => TaskSourceName::PerformanceTimeline,
```
OK, I've filed https://github.com/servo/servo/issues/21527

gterzian commented on components/script/task_queue.rs:
```diff
@@ -0,0 +1,201 @@
+        let throttled_length = self.throttled.borrow().values().fold(0, |acc, queue| acc + queue.len());
+        if throttled_length > 0 {
+            // If we have throttled messages, ensure the select wakes up.
+            let _ = self.wake_up_sender.send(T::wake_up_msg());
```
Good idea, I've moved it to the end of `take_tasks`...

gterzian commented on components/script/task_queue.rs:
```diff
@@ -0,0 +1,201 @@
+        // Always first check for new tasks, but don't reset 'taken_task_counter'.
+        self.process_incoming_tasks();
+        let mut throttled = self.throttled.borrow_mut();
+        let mut throttled_length = throttled.values().fold(0, |acc, queue| acc + queue.len());
```
Ok, it does seem to require a type annotation...

gterzian commented on components/script/task_queue.rs:
```diff
@@ -0,0 +1,201 @@
+        let mut task_source_cycler = task_source_names.iter().cycle();
+        // "being busy", is defined as having more than x tasks for this loop's iteration.
+        // As long as we're not busy, and there are throttled tasks left:
+        while !(max_reached || none_left) {
```
Ok, I've ended up doing it with a `loop` and a match with arms that break if necessary, how does that look?

gterzian commented on components/script/dom/dedicatedworkerglobalscope.rs:
```diff
@@ -153,6 +154,48 @@ pub struct DedicatedWorkerGlobalScope {
+    }
+
+    fn handle_worker_post_event(&self, worker: &TrustedWorkerAddress) {
+        let _ar = AutoWorkerReset::new(&self, worker.clone());
```
I've had it return an `Option<AutoWorkerReset>`, is that ok?

jdm replied:
Oh, yeah, that makes more sense.

gterzian commented:
@jdm Ok I've addressed your comments in two separate commits, matching the original two.

I ended up simplifying the task queue a bit, replacing this `MsgQueue` with just a `VecDequeue` field on the task queue itself, and also not calling `take_tasks` repeatedly like was done previously...

jdm reviewed (changes requested):
Please squash the commits as appropriate!

jdm commented:
Just waiting on squashing into the two commits now.

jdm commented:
@bors-servo delegate=gterzian

bors-servo commented:
:v: @gterzian can now approve this pull request

gterzian commented:
either this or https://github.com/servo/servo/pull/21325 will require a rebase anyhow...

bors-servo commented:
:hourglass: Testing commit 029715aba6f4a4c62b5a2759d47dda57954cc433 with merge 1ee3deea274eeda9a31af3ccd3a9bacf19603ce8...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/5142)

jdm commented:
@bors-servo retry
* #20734

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/12371), [android-x86](http://build.servo.org/builders/android-x86/builds/482), [arm32](http://build.servo.org/builders/arm32/builds/12565), [arm64](http://build.servo.org/builders/arm64/builds/12607), [linux-dev](http://build.servo.org/builders/linux-dev/builds/12669), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/9605), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/9608), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/12696), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/5140), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/9408), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/9451), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/5212), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/5153), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/8071) are reusable. Rebuilding only [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/5142)...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/12371), [android-x86](http://build.servo.org/builders/android-x86/builds/482), [arm32](http://build.servo.org/builders/arm32/builds/12565), [arm64](http://build.servo.org/builders/arm64/builds/12607), [linux-dev](http://build.servo.org/builders/linux-dev/builds/12669), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/9605), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/9608), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/12696), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/5143), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/5140), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/9408), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/9451), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/5212), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/5153), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/8071)
Approved by: jdm
Pushing 1ee3deea274eeda9a31af3ccd3a9bacf19603ce8 to master...
